File: backend/src/durable_objects/rate_limiter.py
"""
Rate Limiter Durable Object for ProtoThrive

Implements distributed rate limiting using Cloudflare Durable Objects
with sliding window algorithm for accurate rate limiting across
all edge locations.

Features:
- Per-user rate limiting
- Per-IP rate limiting
- API key rate limiting
- Sliding window algorithm
- Automatic cleanup of old entries
- DDoS protection
"""

from typing import Dict, List, Optional, Any, Tuple
from datetime import datetime, timedelta
import json
import hashlib
from dataclasses import dataclass, asdict
from enum import Enum

# Cloudflare Durable Object imports
from js import DurableObject, Response, Request


class RateLimitType(Enum):
    """Types of rate limiting"""
    USER = "user"
    IP = "ip"
    API_KEY = "api_key"
    GLOBAL = "global"


@dataclass
class RateLimitConfig:
    """Configuration for rate limits"""
    requests_per_minute: int = 60
    requests_per_hour: int = 1000
    requests_per_day: int = 10000
    burst_size: int = 10
    block_duration_seconds: int = 300  # 5 minutes


@dataclass
class RateLimitWindow:
    """Sliding window data"""
    count: int
    window_start: float
    window_end: float
    blocked_until: Optional[float] = None


class RateLimiterDurableObject(DurableObject):
    """
    Durable Object for distributed rate limiting.

    Each instance handles rate limiting for a specific identifier
    (user, IP, or API key) with consistent state across all edges.
    """

    def __init__(self, state: Any, env: Dict[str, Any]):
        """Initialize rate limiter with state storage"""
        super().__init__(state, env)
        self.state = state
        self.env = env

        # Default configurations per type
        self.configs = {
            RateLimitType.USER: RateLimitConfig(
                requests_per_minute=100,
                requests_per_hour=2000,
                requests_per_day=20000
            ),
            RateLimitType.IP: RateLimitConfig(
                requests_per_minute=60,
                requests_per_hour=1000,
                requests_per_day=10000
            ),
            RateLimitType.API_KEY: RateLimitConfig(
                requests_per_minute=200,
                requests_per_hour=5000,
                requests_per_day=50000
            ),
            RateLimitType.GLOBAL: RateLimitConfig(
                requests_per_minute=1000,
                requests_per_hour=50000,
                requests_per_day=500000
            )
        }

        # Window tracking
        self.windows: Dict[str, List[float]] = {}
        self.blocked: Dict[str, float] = {}

    async def fetch(self, request: Request) -> Response:
        """
        Handle rate limit check request.

        Expected request body:
        {
            "identifier": "user_123",
            "type": "user",
            "increment": true
        }
        """
        try:
            body = await request.json()
            identifier = body.get("identifier")
            limit_type = RateLimitType(body.get("type", "user"))
            increment = body.get("increment", True)

            if not identifier:
                return Response.new(
                    json.dumps({"error": "Identifier required"}),
                    status=400,
                    headers={"Content-Type": "application/json"}
                )

            # Check rate limit
            result = await self.check_rate_limit(
                identifier,
                limit_type,
                increment
            )

            return Response.new(
                json.dumps(asdict(result)),
                status=200,
                headers={"Content-Type": "application/json"}
            )

        except Exception as e:
            return Response.new(
                json.dumps({"error": str(e)}),
                status=500,
                headers={"Content-Type": "application/json"}
            )

    async def check_rate_limit(
        self,
        identifier: str,
        limit_type: RateLimitType,
        increment: bool = True
    ) -> Dict[str, Any]:
        """
        Check if identifier has exceeded rate limit.

        Returns:
            Dictionary with:
            - allowed: bool - Whether request is allowed
            - remaining: int - Requests remaining in window
            - reset_at: float - When window resets (timestamp)
            - retry_after: int - Seconds until retry (if blocked)
        """
        current_time = datetime.utcnow().timestamp()
        config = self.configs[limit_type]

        # Create composite key for storage
        key = f"{limit_type.value}:{identifier}"

        # Check if currently blocked
        if key in self.blocked:
            blocked_until = self.blocked[key]
            if current_time < blocked_until:
                return {
                    "allowed": False,
                    "remaining": 0,
                    "reset_at": blocked_until,
                    "retry_after": int(blocked_until - current_time),
                    "reason": "Rate limit exceeded - blocked"
                }
            else:
                # Block expired, remove it
                del self.blocked[key]

        # Get or create window
        if key not in self.windows:
            self.windows[key] = []

        window = self.windows[key]

        # Clean old entries (older than 1 day)
        cutoff_time = current_time - 86400
        self.windows[key] = [t for t in window if t > cutoff_time]
        window = self.windows[key]

        # Check limits for different time windows
        checks = [
            (60, config.requests_per_minute, "per_minute"),
            (3600, config.requests_per_hour, "per_hour"),
            (86400, config.requests_per_day, "per_day")
        ]

        for window_seconds, limit, limit_name in checks:
            window_start = current_time - window_seconds
            recent_requests = [t for t in window if t > window_start]

            if len(recent_requests) >= limit:
                # Rate limit exceeded - block the identifier
                self.blocked[key] = current_time + config.block_duration_seconds

                # Save state to durable storage
                await self.state.storage.put(f"blocked:{key}", self.blocked[key])

                return {
                    "allowed": False,
                    "remaining": 0,
                    "reset_at": current_time + window_seconds,
                    "retry_after": config.block_duration_seconds,
                    "reason": f"Rate limit exceeded - {limit_name}",
                    "limit": limit,
                    "window": limit_name
                }

        # Check burst protection
        if len(window) > 0:
            recent_burst = [t for t in window if t > current_time - 1]  # Last second
            if len(recent_burst) >= config.burst_size:
                return {
                    "allowed": False,
                    "remaining": 0,
                    "reset_at": current_time + 1,
                    "retry_after": 1,
                    "reason": "Burst limit exceeded"
                }

        # Request allowed
        if increment:
            # Add current request to window
            self.windows[key].append(current_time)

            # Save state periodically (every 10 requests)
            if len(self.windows[key]) % 10 == 0:
                await self.state.storage.put(f"window:{key}", self.windows[key])

        # Calculate remaining requests (use minute window as primary)
        minute_window = [t for t in window if t > current_time - 60]
        remaining = config.requests_per_minute - len(minute_window)

        return {
            "allowed": True,
            "remaining": max(0, remaining),
            "reset_at": current_time + 60,
            "retry_after": 0,
            "requests_in_minute": len(minute_window),
            "limit_per_minute": config.requests_per_minute
        }

    async def reset_limits(self, identifier: str, limit_type: RateLimitType) -> None:
        """Reset rate limits for specific identifier"""
        key = f"{limit_type.value}:{identifier}"

        if key in self.windows:
            del self.windows[key]
        if key in self.blocked:
            del self.blocked[key]

        # Clear from durable storage
        await self.state.storage.delete(f"window:{key}")
        await self.state.storage.delete(f"blocked:{key}")

    async def get_stats(self) -> Dict[str, Any]:
        """Get statistics about current rate limiting state"""
        current_time = datetime.utcnow().timestamp()

        stats = {
            "active_windows": len(self.windows),
            "blocked_identifiers": len(self.blocked),
            "identifiers": {}
        }

        for key, window in self.windows.items():
            limit_type, identifier = key.split(":", 1)
            minute_window = [t for t in window if t > current_time - 60]
            hour_window = [t for t in window if t > current_time - 3600]

            stats["identifiers"][key] = {
                "type": limit_type,
                "identifier": identifier,
                "requests_per_minute": len(minute_window),
                "requests_per_hour": len(hour_window),
                "is_blocked": key in self.blocked
            }

        return stats


class RateLimiterStub:
    """
    Stub implementation for local testing without Durable Objects.

    This provides the same interface but uses in-memory storage.
    """

    def __init__(self):
        self.windows: Dict[str, List[float]] = {}
        self.blocked: Dict[str, float] = {}
        self.configs = {
            RateLimitType.USER: RateLimitConfig(),
            RateLimitType.IP: RateLimitConfig(),
            RateLimitType.API_KEY: RateLimitConfig(),
            RateLimitType.GLOBAL: RateLimitConfig()
        }

    async def check_rate_limit(
        self,
        identifier: str,
        limit_type: str = "user",
        increment: bool = True
    ) -> Dict[str, Any]:
        """Stub implementation of rate limit checking"""
        # For testing, always allow with mock data
        return {
            "allowed": True,
            "remaining": 50,
            "reset_at": datetime.utcnow().timestamp() + 60,
            "retry_after": 0
        }


# Export for use in application
__all__ = [
    'RateLimiterDurableObject',
    'RateLimiterStub',
    'RateLimitType',
    'RateLimitConfig'
]

File: backend/src/durable-objects/RateLimiter.js
/**
 * RateLimiter Durable Object for ProtoThrive
 * Provides distributed, persistent rate limiting across edge locations
 */
export class RateLimiter {
    state;
    env;
    constructor(state, env) {
        this.state = state;
        this.env = env;
    }
    async fetch(request) {
        const url = new URL(request.url);
        const method = request.method;
        if (method === 'POST' && url.pathname === '/check') {
            return this.handleRateLimitCheck(request);
        }
        if (method === 'DELETE' && url.pathname === '/reset') {
            return this.handleReset(request);
        }
        return new Response('Method not allowed', { status: 405 });
    }
    /**
     * Check rate limit for a client
     */
    async handleRateLimitCheck(request) {
        try {
            const body = await request.json();
            const { clientId, config, endpoint } = body;
            const key = endpoint ? `${clientId}:${endpoint}` : clientId;
            const result = await this.checkRateLimit(key, config);
            return new Response(JSON.stringify(result), {
                status: result.allowed ? 200 : 429,
                headers: {
                    'Content-Type': 'application/json',
                    'X-RateLimit-Limit': config.maxRequests.toString(),
                    'X-RateLimit-Remaining': result.remaining.toString(),
                    'X-RateLimit-Reset': new Date(result.resetTime).toISOString(),
                    ...(result.retryAfter && {
                        'Retry-After': Math.ceil(result.retryAfter / 1000).toString()
                    })
                }
            });
        }
        catch (error) {
            console.error('Rate limit check error:', error);
            return new Response(JSON.stringify({
                error: 'Rate limit check failed',
                allowed: false
            }), {
                status: 500,
                headers: { 'Content-Type': 'application/json' }
            });
        }
    }
    /**
     * Reset rate limit for a client
     */
    async handleReset(request) {
        try {
            const body = await request.json();
            const { clientId, endpoint } = body;
            const key = endpoint ? `${clientId}:${endpoint}` : clientId;
            await this.state.storage.delete(key);
            return new Response(JSON.stringify({
                success: true,
                message: 'Rate limit reset successfully'
            }), {
                status: 200,
                headers: { 'Content-Type': 'application/json' }
            });
        }
        catch (error) {
            console.error('Rate limit reset error:', error);
            return new Response(JSON.stringify({
                error: 'Rate limit reset failed',
                success: false
            }), {
                status: 500,
                headers: { 'Content-Type': 'application/json' }
            });
        }
    }
    /**
     * Core rate limiting logic with token bucket algorithm
     */
    async checkRateLimit(key, config) {
        const now = Date.now();
        const windowMs = config.windowMs;
        const maxRequests = config.maxRequests;
        const burstLimit = config.burstLimit || Math.floor(maxRequests * 1.5);
        // Get current state
        const storedState = await this.state.storage.get(key);
        let state;
        if (!storedState || now > storedState.resetTime) {
            // Initialize or reset window
            state = {
                count: 1,
                resetTime: now + windowMs,
                burst: 1,
                lastRequest: now
            };
        }
        else {
            // Update existing state
            const timeSinceLastRequest = now - storedState.lastRequest;
            const tokensToAdd = Math.floor(timeSinceLastRequest * maxRequests / windowMs);
            state = {
                count: storedState.count + 1,
                resetTime: storedState.resetTime,
                burst: Math.min(burstLimit, storedState.burst + tokensToAdd),
                lastRequest: now
            };
        }
        // Check rate limits
        const allowed = state.count <= maxRequests && state.burst > 0;
        if (allowed) {
            state.burst = Math.max(0, state.burst - 1);
            // Store updated state
            await this.state.storage.put(key, state);
        }
        const remaining = Math.max(0, maxRequests - state.count);
        const retryAfter = allowed ? undefined : state.resetTime - now;
        return {
            allowed,
            remaining,
            resetTime: state.resetTime,
            retryAfter
        };
    }
    /**
     * Clean up expired entries periodically
     */
    async cleanup() {
        const now = Date.now();
        const allEntries = await this.state.storage.list();
        const expiredKeys = [];
        for (const [key, state] of allEntries) {
            if (now > state.resetTime + 60000) { // Grace period of 1 minute
                expiredKeys.push(key);
            }
        }
        if (expiredKeys.length > 0) {
            await this.state.storage.delete(expiredKeys);
            console.log(`Cleaned up ${expiredKeys.length} expired rate limit entries`);
        }
    }
}
/**
 * Rate limiting configurations for different endpoints
 */
export const RATE_LIMIT_CONFIGS = {
    // Authentication endpoints - stricter limits
    '/api/auth/login': {
        maxRequests: 5,
        windowMs: 15 * 60 * 1000, // 15 minutes
        burstLimit: 3
    },
    '/api/auth/register': {
        maxRequests: 3,
        windowMs: 60 * 60 * 1000, // 1 hour
        burstLimit: 2
    },
    // API endpoints - standard limits
    '/api/roadmaps': {
        maxRequests: 100,
        windowMs: 60 * 1000, // 1 minute
        burstLimit: 150
    },
    '/api/snippets': {
        maxRequests: 50,
        windowMs: 60 * 1000, // 1 minute
        burstLimit: 75
    },
    // AI agent endpoints - resource intensive
    '/api/ai/orchestrate': {
        maxRequests: 10,
        windowMs: 60 * 1000, // 1 minute
        burstLimit: 5
    },
    // Default configuration
    'default': {
        maxRequests: 1000,
        windowMs: 60 * 1000, // 1 minute
        burstLimit: 1500
    }
};
